parachain/pallets/control/src/lib.rs

```diff
@@ -170,6 +209,12 @@ pub mod pallet {
 				location
 			);
 
+			let agent_owner = Self::agent_account_id(agent_id)?;
```

Comment:
Just to clarify, we need to convert `origin_location` into a signed sovereign account. So `MultiLocation` to  `AccountId`.

Why? because this allows the origin top up their sovereign account on BridgeHub using normal XCM transfers/teleports. This allows them to pay for the deposit when calling `create_agent`.

For example,  see how pallet-xcm uses this converter:
https://github.com/paritytech/polkadot-sdk/blob/430edd75351e82ea3f814c05fbbbe0cb860e6d27/polkadot/xcm/pallet-xcm/src/lib.rs#L251

https://github.com/paritytech/polkadot-sdk/blob/430edd75351e82ea3f814c05fbbbe0cb860e6d27/polkadot/xcm/pallet-xcm/src/lib.rs#L1758

Reply:
> need to convert origin_location into a signed sovereign account

 I would assume the signed sovereign you mentioned above refer in particular to [AccountId32Aliases](https://github.com/paritytech/polkadot-sdk/blob/430edd75351e82ea3f814c05fbbbe0cb860e6d27/cumulus/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/xcm_config.rs#L85C1-L85C1)  which applies to `AccountId32` only.

AFAIK in our design there is no limitation for the xcm origin, actually there are several other use cases in unit tests which allows from [PalletInstance](https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/parachain/pallets/control/src/mock.rs#L100-L106) or [AccountKey20](https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/parachain/pallets/control/src/mock.rs#L119-L128),  would assume these are all valid use cases that we need to support.

>  allows the origin top up their sovereign account on BridgeHub using normal XCM transfers/teleports.

 I would assume actually the derivated sovereign account is keyless and does not require that particular signed origin to top up it, actually it's the dest account anyone can fund it with a xcm transfer(e.g. check how we fund the gateway sovereign from relaychain in https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/web/packages/test/scripts/configure-bridgehub.sh#L42)





Btw:  I would assume does not require that particular signed origin to top up the derivated sovereign account, could be from arbitrary account as we initial fund the  

https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/web/packages/test/scripts/configure-bridgehub.sh#L39

Reply:
Since `agent_id` is also a unique identifier derivated from `MultiLocation` with [AgentHashedDescription](https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/parachain/pallets/control/src/lib.rs#L66), so I would like to reuse it as the derivated sovereign account.

Reply:
https://linear.app/snowfork/issue/SNO-625#comment-c20c5032

Reply:
> > need to convert origin_location into a signed sovereign account
> 

Oops, I meant convert into a just a sovereign account. Is keyless as you say.

> Since `agent_id` is already a unique identifier derivated from `MultiLocation` with [AgentHashedDescription](https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/parachain/pallets/control/src/lib.rs#L66), so I would prefer to reuse it rather than introducing another origin converter.

Agents, agent addresses, and agent ids are mainly for use on the Ethereum side of the bridge, where using `MultiLocation` is not possible. On the Polkadot side of the bridge we shouldn't be complicating the matter of identity by using agent ids to refer to Polkadot origins. It also introduces an unnecessary 2-step conversion to transform origins to sovereign accounts: `MultiLocation -> AgentID -> SovereignAccount`.

Reply:
> I would assume actually the derivated sovereign account is keyless and does not require some particular signed origin to fund it, actually it's a dest account anyone can fund through xcm transfer(e.g. check how we [fund](https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/web/packages/test/scripts/configure-bridgehub.sh#L42) the gateway sovereign from relaychain).

Sure, anyone can fund the sovereign account derived from the agent, but no one is able to withdraw funds from the account, unless we add a special extrinsic to allow it. 

If we just create a sovereign account based on the `MultiLocation`, then the origin is able to withdraw using XCM without us adding any special logic

Reply:
> Agents, agent addresses, and agent ids are mainly for use on the Ethereum side of the bridge, where using MultiLocation is not possible. On the Polkadot side of the bridge we shouldn't be complicating the matter of identity by using agent ids to refer to Polkadot origins

Yes, `agent_address` is used on the Ethereum side to represent the agent, but I would assume `agent_id` is not, it's derivated through [AgentHashedDescription](https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/parachain/pallets/control/src/lib.rs#L66) purely from the `Multilocation`, check the internal implementation [here](https://github.com/paritytech/polkadot-sdk/blob/43c0c09bcb9a4db0856598dd07357c0b7292bb84/polkadot/xcm/xcm-builder/src/location_conversion.rs#L120) nothing to do with the Ethereum side things.

And in current implementation the sovereign account is identical as `agent_id`(both 32 bytes).

Reply:
>  If we just create a sovereign account based on the MultiLocation, then the origin is able to withdraw using XCM without us adding any special logic

AFAIK `bridgeHub` is configured not allow to send arbitray xcm messages through `xcmPallet`, check the config [here](https://github.com/Snowfork/cumulus/blob/38c2e26a4ead087d0837b5abdc3cdc3e0e6e0d8e/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/xcm_config.rs#L311-L312)
which disable the send extrinsic and config [here](https://github.com/Snowfork/cumulus/blob/38c2e26a4ead087d0837b5abdc3cdc3e0e6e0d8e/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/xcm_config.rs#L315) which disable the execute extrinsic.

So I think we do need some special logic(a custom call) to withdraw from sovereign anyway.

Reply:
> So I think we do need some special logic(a custom call) allow to withdraw from sovereign anyway.

Sure, but only bridge hub governance to withdraw its funds. Other origins on Polkadot can withdraw funds from their sovereign accounts without using any special logic, as long as we derive the sovereign account from the MultiLocation.

I really would like to follow the convention that other parachains are using to create sovereign accounts, which is to use some [converter](https://github.com/Snowfork/cumulus/blob/38c2e26a4ead087d0837b5abdc3cdc3e0e6e0d8e/parachains/runtimes/bridge-hubs/bridge-hub-rococo/src/xcm_config.rs#L94) already available in the polkadot xcm libs. It's going to reduce surprises in future.

Reply:
Also in the latest Polkadot code, you can use this instead of `LocationToAccountId`: https://github.com/paritytech/polkadot-sdk/blob/73915d3459a0723e2ca80a5ca5d0a11a3736b18d/polkadot/xcm/xcm-builder/src/location_conversion.rs#L121

Reply:
OK, I took a deeper look at the implementation of [AgentHashedDescription](https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/parachain/pallets/control/src/lib.rs#L66) and I see that it basically is what we need to use to convert MultiLocations.

As you say, we cannot use `LocationToAccountId`, since it doesn't support converting AccountId32 junctions.

So I was technically wrong on a few points.

Reply:
See https://github.com/Snowfork/snowbridge/pull/940/files#r1310744745, we can use this converter in Polkadot. It already uses some of the same code as [AgentHashedDescription](https://github.com/Snowfork/snowbridge/blob/1201e1db5aa6ae580123a745bc7387c2019c4321/parachain/pallets/control/src/lib.rs#L66), but returns `AccountId` instead of `H256`.